compiler/optimizer/VirtualGuardHeadMerger.cpp

```diff
@@ -467,6 +478,20 @@ int32_t TR_VirtualGuardHeadMerger::perform() {
             if (guard2->getBranchDestination() != guard1->getBranchDestination())
                guard2Block->changeBranchDestination(guard1->getBranchDestination(), cfg);
 
+            // changeBranchDestination removes unreachable blocks after the removal of the old edge from guard2 to cold2
+            // In rare cases, guard2Block could become unreachable and removed. Normally, if all relevant blocks are removed,
```

Comment:
> In rare cases, `guard2Block` could become unreachable and removed.

I don't understand how this is possible. Before `changeBranchDestination()`, `guard2Block` should be reachable to begin with. (With the changes in this PR, the loop has even already checked to make sure, though I don't understand why we should need to do that either.) Since `guard2Block` is reachable, there's a path to it that doesn't pass through it, and that path is unaffected by changes to edges outgoing from `guard2Block`

Reply:
> Before changeBranchDestination(), guard2Block should be reachable to begin with. 

Correct, guard2Block is reachable before `changeBranchDestination` but can become unreachable after `changeBranchDestination` in rare cases. `changeBranchDestination` removes the old edge from guard2 to cold2 and also subsequent unreachable blocks caused by the removal of the old edge [1]. 

https://github.com/eclipse-openj9/openj9/issues/18873#issuecomment-2018929069 gives a specific example of how guard2Block becomes unreachable after `changeBranchDestination` and how `moveBlockAfterDest` later messed up  the CFG. 

> the loop has even already checked to make sure, though I don't understand why we should need to do that either.

The check in the beginning of the loop is just for precaution, which I was debating if it was really necessary. It can be removed if it is concluded not necessary. 

However the check in the beginning of the loop will surely not prevent the problem if `changeBranchDestination` causes guard2 block being removed because `moveBlockAfterDest` could be called after `changeBranchDestination` before it gets to the beginning of the loop. The second check after `changeBranchDestination`  is the required fix here


[1]
https://github.com/eclipse/omr/blob/6391cf4bb4f415ba5f6116311dfd3c94d2c92cd7/compiler/il/OMRBlock.cpp#L752
https://github.com/eclipse/omr/blob/6391cf4bb4f415ba5f6116311dfd3c94d2c92cd7/compiler/infra/OMRCfg.cpp#L953-L965

Reply:
https://github.com/eclipse-openj9/openj9/issues/18873#issuecomment-2025513238 contains a more detailed log that shows the same issue

Reply:
OK, so there were unreachable blocks left in the CFG from a previous `changeBranchDestination()` in the same pass of VGHM. I thought that we had an invariant that the CFG consists only of reachable blocks at all times. However, after discussing with Annabelle I see that removing edges can sometimes leave unreachable blocks in the CFG, and they will be removed at the end of the opt pass but not necessarily beforehand. In particular, after deleting an edge, `!nodeIsRemoved()` doesn't necessarily mean that a block is reachable

So with that, the logic change here looks good, including the check at the top of the loop :+1:

I think some parts of the comment are misleading though

> `changeBranchDestination` removes unreachable blocks after the removal of the old edge

To me this sounds like `changeBranchDestination()` removes all of the newly unreachable blocks, which is what I thought it did, but the bug fixed here hinges on it sometimes making blocks unreachable without also removing them (and other times actually removing unreachable blocks)

> `guard2Block` could become unreachable and removed.

It can be removed, but only if it's already unreachable

Here's a suggestion for the wording, in case you like it:

> `changeBranchDestination` *tries to remove* unreachable blocks after the removal of the old edge from `guard2` to `cold2`. *Rarely, a previous transformation in the same pass could have made `guard2Block` unreachable without removing it, in which case it might have been removed just now.* Normally, [...]

Reply:
Thank you for the suggestion, which definitely reflects more accurately on what happens in this case!

Addressed in 3f7c770. Ready for another review